src/flamenco/vm/syscall/fd_vm_syscall_util.c

```diff
@@ -380,98 +380,99 @@ fd_vm_syscall_sol_memcpy( /**/            void *  _vm,
+      dst_bytes_rem_in_cur_region = fd_ulong_min( sz, fd_ulong_sat_sub( vm->region_st_sz[dst_region], dst_offset ) );
     }
 
-    ulong src_region              = src_vaddr >> 32;
-    ulong src_offset              = src_vaddr & 0xffffffffUL;
-    ulong src_region_idx          = 0UL;
-    ulong src_bytes_in_cur_region = sz;
-    uchar * src_haddr             = NULL;
-    if( src_region==4UL ) {
-      src_region_idx          = fd_vm_get_input_mem_region_idx( vm, src_offset );
-      src_haddr               = (uchar*)(vm->input_mem_regions[ src_region_idx ].haddr + src_offset - vm->input_mem_regions[ src_region_idx ].vaddr_offset);
-      src_bytes_in_cur_region = fd_ulong_min( sz, fd_ulong_sat_sub( vm->input_mem_regions[ src_region_idx ].region_sz, 
-                                                                    ((ulong)src_haddr - vm->input_mem_regions[ src_region_idx ].haddr) ) );
-      if( FD_UNLIKELY( src_region_idx+1UL==vm->input_mem_regions_cnt && src_bytes_in_cur_region<sz ) ) {
-        *_ret = 1;
-        return FD_VM_ERR_ABORT;
-      }
-
+    ulong   src_region                  = fd_ulong_min( src_vaddr >> 32, 5UL );
+    uchar   src_is_input_mem_region     = ( src_region==4UL );
+    ulong   src_offset                  = src_vaddr & 0xffffffffUL;
+    ulong   src_region_idx              = 0UL;
+    ulong   src_bytes_rem_in_cur_region;
+    uchar * src_haddr;
+    if( src_is_input_mem_region ) {
+      FD_VM_MEM_HADDR_AND_REGION_IDX_FROM_INPUT_REGION_UNCHECKED( vm, src_offset, src_region_idx, src_haddr );
+      src_bytes_rem_in_cur_region = fd_ulong_sat_sub( vm->input_mem_regions[ src_region_idx ].region_sz, (src_offset - vm->input_mem_regions[ src_region_idx ].vaddr_offset) );
     } else {
-      src_haddr           = (uchar *)FD_VM_MEM_SLICE_HADDR_LD( vm, src_vaddr, FD_VM_ALIGN_RUST_U8, sz );
-    }
-
-    /* Do a normal memcpy if regions do not overlap */
-    if( FD_LIKELY( src_bytes_in_cur_region==dst_bytes_in_cur_region && src_bytes_in_cur_region==sz ) ) {
-      fd_memcpy( dst_haddr, src_haddr, sz );
-      *_ret = 0;
-      return FD_VM_SUCCESS;
+      src_haddr = (uchar*)FD_VM_MEM_HADDR_LD_FAST( vm, src_vaddr );
+      src_bytes_rem_in_cur_region = fd_ulong_min( sz, fd_ulong_sat_sub( vm->region_ld_sz[src_region], src_offset ) );
     }
   
-    /* Case where the operation spans multiple regions. Copy over the bytes
-       from each region while iterating to the next one. */
-    /* TODO: An optimization would be to memcpy chunks at once */
-    ulong dst_idx = 0UL;
-    ulong src_idx = 0UL;
-    for( ulong i=0UL; i<sz; i++ ) {
-      if( FD_UNLIKELY( !dst_bytes_in_cur_region ) ) {
-        /* Go to next one */
-        if( FD_UNLIKELY( ++dst_region_idx>=vm->input_mem_regions_cnt ) ) {
-          *_ret = 1;
-          return FD_VM_ERR_ABORT;
+    /* Copy over the bytes from each region in chunks. */
+    while( sz>0UL ) {
+      if( dst_is_input_mem_region ) {
```

Comment:
reorder these s.t. we firs tcheck for if there are bytes in the region. Do the same for the src buffer(s)